# Business Problem

Our Client is a medical company situated in Nairobi,Kenya. Based on the recent outbreak of the Corona Virus and the recent developments which led to the discovery of the vaccine. They want you to use past data on H1N1 influenza virus which is on different people’s backgrounds, opinions, and health behaviors to come up with an algoritm that can predict whether someone got the H1N1 vaccine and with this tehy will get a better understanding of how these characteristics have been associated with personal vaccination patterns may provide guidance for future public health efforts.

# Labels

1. h1n1_vaccine - Whether respondent received H1N1 flu vaccine.

2. seasonal_vaccine - Whether respondent received seasonal flu vaccine.

Both are binary variables: 0 = No; 1 = Yes. Some respondents didn't get either vaccine, others got only one, and some got both. This is formulated as a multilabel (and not multiclass) problem.

# The features in this dataset


You are provided a dataset with 36 columns. The first column respondent_id is a unique and random identifier. The remaining 35 features are described below.

For all binary variables: 0 = No; 1 = Yes.

1. h1n1_concern - Level of concern about the H1N1 flu.
   0 = Not at all concerned; 1 = Not very concerned; 2 = Somewhat concerned; 3 = Very concerned.

2. h1n1_knowledge - Level of knowledge about H1N1 flu.
   0 = No knowledge; 1 = A little knowledge; 2 = A lot of knowledge.

3. behavioral_antiviral_meds - Has taken antiviral medications. (binary)
 
4. behavioral_avoidance - Has avoided close contact with others with flu-like symptoms. (binary)

5. behavioral_face_mask - Has bought a face mask. (binary)

6. behavioral_wash_hands - Has frequently washed hands or used hand sanitizer. (binary)

7. behavioral_large_gatherings - Has reduced time at large gatherings. (binary)

8. behavioral_outside_home - Has reduced contact with people outside of own household. (binary)

9. behavioral_touch_face - Has avoided touching eyes, nose, or mouth. (binary)

10. doctor_recc_h1n1 - H1N1 flu vaccine was recommended by doctor. (binary)

11. doctor_recc_seasonal - Seasonal flu vaccine was recommended by doctor. (binary)

12. chronic_med_condition - Has any of the following chronic medical conditions: asthma or an other lung condition, diabetes, a heart condition, a kidney   condition, sickle cell anemia or other anemia, a neurological or neuromuscular condition, a liver condition, or a weakened immune system caused by a chronic illness or by medicines taken for a chronic illness. (binary)

13. child_under_6_months - Has regular close contact with a child under the age of six months. (binary)

14. health_worker - Is a healthcare worker. (binary)

15. health_insurance - Has health insurance. (binary)

16. opinion_h1n1_vacc_effective - Respondent's opinion about H1N1 vaccine effectiveness.
    1 = Not at all effective; 2 = Not very effective; 3 = Don't know; 4 = Somewhat effective; 5 = Very effective.

17. opinion_h1n1_risk - Respondent's opinion about risk of getting sick with H1N1 flu without vaccine.
    1 = Very Low; 2 = Somewhat low; 3 = Don't know; 4 = Somewhat high; 5 = Very high.

18. opinion_h1n1_sick_from_vacc - Respondent's worry of getting sick from taking H1N1 vaccine.
    1 = Not at all worried; 2 = Not very worried; 3 = Don't know; 4 = Somewhat worried; 5 = Very worried.

19. opinion_seas_vacc_effective - Respondent's opinion about seasonal flu vaccine effectiveness.
    1 = Not at all effective; 2 = Not very effective; 3 = Don't know; 4 = Somewhat effective; 5 = Very effective.

20. opinion_seas_risk - Respondent's opinion about risk of getting sick with seasonal flu without vaccine.
    1 = Very Low; 2 = Somewhat low; 3 = Don't know; 4 = Somewhat high; 5 = Very high.

21. opinion_seas_sick_from_vacc - Respondent's worry of getting sick from taking seasonal flu vaccine.
    1 = Not at all worried; 2 = Not very worried; 3 = Don't know; 4 = Somewhat worried; 5 = Very worried.

22. age_group - Age group of respondent.

23. education - Self-reported education level.

24. race - Race of respondent.

25. sex - Sex of respondent.

26. income_poverty - Household annual income of respondent with respect to 2008 Census poverty thresholds.

27. marital_status - Marital status of respondent.

28. rent_or_own - Housing situation of respondent.

29. employment_status - Employment status of respondent.

30. hhs_geo_region - Respondent's residence using a 10-region geographic classification defined by the U.S. Dept. of Health and Human Services. Values are  represented as short random character strings.

31. census_msa - Respondent's residence within metropolitan statistical areas (MSA) as defined by the U.S. Census.

32. household_adults - Number of other adults in household, top-coded to 3.

33. household_children - Number of children in household, top-coded to 3.

34. employment_industry - Type of industry respondent is employed in. Values are represented as short random character strings.

35. employment_occupation - Type of occupation of respondent. Values are represented as short random character strings.

# Load The Data

Below, i will import the relevant libraries for the data analysis and model. I will also load the data.

In [1]:
# Load the relevant libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split


The data provided has already been split into the train and test dataset. The target labels for the training data are also stored in a separate dataframes. Therefore i will load the datasets into different data frame with the Features dataframe containing the training data features/independent variables while the Target dataframe will contain our training data target/dependent variable.

In [2]:
# Load the data
Features = pd.read_csv("Data/training_set_features.csv")
Target = pd.read_csv("Data/training_set_labels.csv")

In [3]:
# Load the first 5 rows of the data
Features.head()

respondent_id  h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
0              0           1.0             0.0                        0.0   
1              1           3.0             2.0                        0.0   
2              2           1.0             1.0                        0.0   
3              3           1.0             1.0                        0.0   
4              4           2.0             1.0                        0.0   

   behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
0                   0.0                   0.0                    0.0   
1                   1.0                   0.0                    1.0   
2                   1.0                   0.0                    0.0   
3                   1.0                   0.0                    1.0   
4                   1.0                   0.0                    1.0   

   behavioral_large_gatherings  behavioral_outside_home  \
0                          0.0                      1.0   
1                          0.0                      1.0   
2                          0.0                      0.0   
3                          1.0                      0.0   
4                          1.0                      0.0   

   behavioral_touch_face  ...             income_poverty  marital_status  \
0                    1.0  ...              Below Poverty     Not Married   
1                    1.0  ...              Below Poverty     Not Married   
2                    0.0  ...  <= $75,000, Above Poverty     Not Married   
3                    0.0  ...              Below Poverty     Not Married   
4                    1.0  ...  <= $75,000, Above Poverty         Married   

   rent_or_own   employment_status  hhs_geo_region                census_msa  \
0          Own  Not in Labor Force        oxchjgsf                   Non-MSA   
1         Rent            Employed        bhuqouqj  MSA, Not Principle  City   
2          Own            Employed        qufhixun  MSA, Not Principle  City   
3         Rent  Not in Labor Force        lrircsnp       MSA, Principle City   
4          Own            Employed        qufhixun  MSA, Not Principle  City   

   household_adults  household_children  employment_industry  \
0               0.0                 0.0                  NaN   
1               0.0                 0.0             pxcmvdjn   
2               2.0                 0.0             rucpziij   
3               0.0                 0.0                  NaN   
4               1.0                 0.0             wxleyezf   

   employment_occupation  
0                    NaN  
1               xgwztkwe  
2               xtkaffoo  
3                    NaN  
4               emcorrxb  

[5 rows x 36 columns]

In [4]:
#set the respondent_id as the index
Features.set_index('respondent_id', inplace=True)
Features.head()

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
0                       1.0             0.0                        0.0   
1                       3.0             2.0                        0.0   
2                       1.0             1.0                        0.0   
3                       1.0             1.0                        0.0   
4                       2.0             1.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
0                               0.0                   0.0   
1                               1.0                   0.0   
2                               1.0                   0.0   
3                               1.0                   0.0   
4                               1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
0                                0.0                          0.0   
1                                1.0                          0.0   
2                                0.0                          0.0   
3                                1.0                          1.0   
4                                1.0                          1.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
0                                  1.0                    1.0   
1                                  1.0                    1.0   
2                                  0.0                    0.0   
3                                  0.0                    0.0   
4                                  0.0                    1.0   

               doctor_recc_h1n1  ...             income_poverty  \
respondent_id                    ...                              
0                           0.0  ...              Below Poverty   
1                           0.0  ...              Below Poverty   
2                           NaN  ...  <= $75,000, Above Poverty   
3                           0.0  ...              Below Poverty   
4                           0.0  ...  <= $75,000, Above Poverty   

               marital_status  rent_or_own   employment_status  \
respondent_id                                                    
0                 Not Married          Own  Not in Labor Force   
1                 Not Married         Rent            Employed   
2                 Not Married          Own            Employed   
3                 Not Married         Rent  Not in Labor Force   
4                     Married          Own            Employed   

               hhs_geo_region                census_msa  household_adults  \
respondent_id                                                               
0                    oxchjgsf                   Non-MSA               0.0   
1                    bhuqouqj  MSA, Not Principle  City               0.0   
2                    qufhixun  MSA, Not Principle  City               2.0   
3                    lrircsnp       MSA, Principle City               0.0   
4                    qufhixun  MSA, Not Principle  City               1.0   

               household_children  employment_industry  employment_occupation  
respondent_id                                                                  
0                             0.0                  NaN                    NaN  
1                             0.0             pxcmvdjn               xgwztkwe  
2                             0.0             rucpziij               xtkaffoo  
3                             0.0                  NaN                    NaN  
4                             0.0             wxleyezf               emcorrxb  

[5 rows x 35 columns]

In [5]:
# Load the first 5 rows of the target data
Target.head()

respondent_id  h1n1_vaccine  seasonal_vaccine
0              0             0                 0
1              1             0                 1
2              2             0                 0
3              3             0                 1
4              4             0                 0

In [6]:
# set the respondent_id as the index
Target.set_index('respondent_id', inplace=True)
Target.head()

h1n1_vaccine  seasonal_vaccine
respondent_id                                
0                         0                 0
1                         0                 1
2                         0                 0
3                         0                 1
4                         0                 0

In [7]:
# Load the summary of the Features Dataframe
Features.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26707 entries, 0 to 26706
Data columns (total 35 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   h1n1_concern                 26615 non-null  float64
 1   h1n1_knowledge               26591 non-null  float64
 2   behavioral_antiviral_meds    26636 non-null  float64
 3   behavioral_avoidance         26499 non-null  float64
 4   behavioral_face_mask         26688 non-null  float64
 5   behavioral_wash_hands        26665 non-null  float64
 6   behavioral_large_gatherings  26620 non-null  float64
 7   behavioral_outside_home      26625 non-null  float64
 8   behavioral_touch_face        26579 non-null  float64
 9   doctor_recc_h1n1             24547 non-null  float64
 10  doctor_recc_seasonal         24547 non-null  float64
 11  chronic_med_condition        25736 non-null  float64
 12  child_under_6_months         25887 non-null  float64
 13  health_worker   

In [8]:
# Load the summary of the Target Dataframe
Target.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 26707 entries, 0 to 26706
Data columns (total 2 columns):
 #   Column            Non-Null Count  Dtype
---  ------            --------------  -----
 0   h1n1_vaccine      26707 non-null  int64
 1   seasonal_vaccine  26707 non-null  int64
dtypes: int64(2)
memory usage: 625.9 KB


Based on the summary on the two dataframes provided above we can tell that:
1. The features dataframe is made up of 35 columns, most of the columns are of the int and float datatype with some are of the object datatype. The        dataframe has 26707 entiries in each column with some columns contaning null entries.
2. The Target dataframe is made up of 2 columns both of the integer datatype. There are no null entries.

Based on the fact that our business problem is more targeted towards the H1N1 influenza virus vaccine, i will go ahead and drop the seasonal_vaccine target label from the Target Dataframe.

In [9]:
# Drop the 'seasonal_vaccine' column
Target.drop('seasonal_vaccine', axis=1, inplace=True)

In [10]:
# Load the first five columns of the Target DataFrame
Target.head()

h1n1_vaccine
respondent_id              
0                         0
1                         0
2                         0
3                         0
4                         0

Below, i will generate the descriptive statistics of the Features dataframe.

In [11]:
# Load the descriptive statistics
Features.describe()

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
count  26615.000000    26591.000000               26636.000000   
mean       1.618486        1.262532                   0.048844   
std        0.910311        0.618149                   0.215545   
min        0.000000        0.000000                   0.000000   
25%        1.000000        1.000000                   0.000000   
50%        2.000000        1.000000                   0.000000   
75%        2.000000        2.000000                   0.000000   
max        3.000000        2.000000                   1.000000   

       behavioral_avoidance  behavioral_face_mask  behavioral_wash_hands  \
count          26499.000000          26688.000000           26665.000000   
mean               0.725612              0.068982               0.825614   
std                0.446214              0.253429               0.379448   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000               1.000000   
50%                1.000000              0.000000               1.000000   
75%                1.000000              0.000000               1.000000   
max                1.000000              1.000000               1.000000   

       behavioral_large_gatherings  behavioral_outside_home  \
count                  26620.00000             26625.000000   
mean                       0.35864                 0.337315   
std                        0.47961                 0.472802   
min                        0.00000                 0.000000   
25%                        0.00000                 0.000000   
50%                        0.00000                 0.000000   
75%                        1.00000                 1.000000   
max                        1.00000                 1.000000   

       behavioral_touch_face  doctor_recc_h1n1  ...  health_worker  \
count           26579.000000      24547.000000  ...   25903.000000   
mean                0.677264          0.220312  ...       0.111918   
std                 0.467531          0.414466  ...       0.315271   
min                 0.000000          0.000000  ...       0.000000   
25%                 0.000000          0.000000  ...       0.000000   
50%                 1.000000          0.000000  ...       0.000000   
75%                 1.000000          0.000000  ...       0.000000   
max                 1.000000          1.000000  ...       1.000000   

       health_insurance  opinion_h1n1_vacc_effective  opinion_h1n1_risk  \
count       14433.00000                 26316.000000       26319.000000   
mean            0.87972                     3.850623           2.342566   
std             0.32530                     1.007436           1.285539   
min             0.00000                     1.000000           1.000000   
25%             1.00000                     3.000000           1.000000   
50%             1.00000                     4.000000           2.000000   
75%             1.00000                     5.000000           4.000000   
max             1.00000                     5.000000           5.000000   

       opinion_h1n1_sick_from_vacc  opinion_seas_vacc_effective  \
count                 26312.000000                 26245.000000   
mean                      2.357670                     4.025986   
std                       1.362766                     1.086565   
min                       1.000000                     1.000000   
25%                       1.000000                     4.000000   
50%                       2.000000                     4.000000   
75%                       4.000000                     5.000000   
max                       5.000000                     5.000000   

       opinion_seas_risk  opinion_seas_sick_from_vacc  household_adults  \
count       26193.000000                 26170.000000      26458.000000   
mean            2.719162                     2.118112          0.886499   
std             1.38505

Based, on the output above we get a statistical summary on the numerical datatype columns in our dataframe providing us with information on the measures such as the columns mean, standard deviation, interquartile range etc. It also provides the minimum and maximum values in each column which will be helpful in identfying the outliers.

# Data Cleaning
In this section we get to clean our data and prepare the for the exploratory analysis and modeling.

## Checking For Null Values
Below i will check the dataframes for null entries and find ways of handling the null entries.

In [12]:
# check the target dataframe for null entries
Target.isna().sum()

h1n1_vaccine    0
dtype: int64

Based on the output above, the Target Dataframe does not contain null entries.

In [13]:
# check the Features dataframe for null entries
Features.isna().sum()

h1n1_concern                      92
h1n1_knowledge                   116
behavioral_antiviral_meds         71
behavioral_avoidance             208
behavioral_face_mask              19
behavioral_wash_hands             42
behavioral_large_gatherings       87
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_h1n1                2160
doctor_recc_seasonal            2160
chronic_med_condition            971
child_under_6_months             820
health_worker                    804
health_insurance               12274
opinion_h1n1_vacc_effective      391
opinion_h1n1_risk                388
opinion_h1n1_sick_from_vacc      395
opinion_seas_vacc_effective      462
opinion_seas_risk                514
opinion_seas_sick_from_vacc      537
age_group                          0
education                       1407
race                               0
sex                                0
income_poverty                  4423
marital_status                  1408
r

Based on the output, most of the columns in the Features data frame contain null entries. Below i will do investigate what percentage of the entire dataset does the null entries represent.

In [14]:
# Find what percentage the Null values represent
Features.isnull().sum() * 100 / len(Features)

h1n1_concern                    0.344479
h1n1_knowledge                  0.434343
behavioral_antiviral_meds       0.265848
behavioral_avoidance            0.778822
behavioral_face_mask            0.071142
behavioral_wash_hands           0.157262
behavioral_large_gatherings     0.325757
behavioral_outside_home         0.307036
behavioral_touch_face           0.479275
doctor_recc_h1n1                8.087767
doctor_recc_seasonal            8.087767
chronic_med_condition           3.635751
child_under_6_months            3.070356
health_worker                   3.010447
health_insurance               45.957989
opinion_h1n1_vacc_effective     1.464036
opinion_h1n1_risk               1.452803
opinion_h1n1_sick_from_vacc     1.479013
opinion_seas_vacc_effective     1.729884
opinion_seas_risk               1.924589
opinion_seas_sick_from_vacc     2.010709
age_group                       0.000000
education                       5.268282
race                            0.000000
sex             

Based on the output, we see that most of the columns contain missing values that represent a percentage that is less than 10 which is really small. However, 4 columns have percentage that is greater than 10 with 'employment_occupation' having a percentage of 50% which is the highest followed closely by 'employment_industry' with 49.91%.

Next i will split the features dataframe into numeric datatype and categorical datatype

In [15]:
numeric_features = Features.columns[Features.dtypes != "object"]

numeric_features = Features[numeric_features]
numeric_features

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
0                       1.0             0.0                        0.0   
1                       3.0             2.0                        0.0   
2                       1.0             1.0                        0.0   
3                       1.0             1.0                        0.0   
4                       2.0             1.0                        0.0   
...                     ...             ...                        ...   
26702                   2.0             0.0                        0.0   
26703                   1.0             2.0                        0.0   
26704                   2.0             2.0                        0.0   
26705                   1.0             1.0                        0.0   
26706                   0.0             0.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
0                               0.0                   0.0   
1                               1.0                   0.0   
2                               1.0                   0.0   
3                               1.0                   0.0   
4                               1.0                   0.0   
...                             ...                   ...   
26702                           1.0                   0.0   
26703                           1.0                   0.0   
26704                           1.0                   1.0   
26705                           0.0                   0.0   
26706                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
0                                0.0                          0.0   
1                                1.0                          0.0   
2                                0.0                          0.0   
3                                1.0                          1.0   
4                                1.0                          1.0   
...                              ...                          ...   
26702                            0.0                          0.0   
26703                            1.0                          0.0   
26704                            1.0                          1.0   
26705                            0.0                          0.0   
26706                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
0                                  1.0                    1.0   
1                                  1.0                    1.0   
2                                  0.0                    0.0   
3                                  0.0                    0.0   
4                                  0.0                    1.0   
...                                ...                    ...   
26702                              1.0                    0.0   
26703                              0.0                    0.0   
26704                              0.0                    1.0   
26705                              0.0                    NaN   
26706                              0.0                    0.0   

               doctor_recc_h1n1  ...  health_worker  health_insurance  \
respondent_id                    ...                                    
0                           0.0  ...            0.0               1.0   
1                           0.0  ...            0.0               1.0   
2                           NaN  ...            0.0               NaN   
3                           0.0  ...            0.0               NaN   
4                           0.0  ...            0.0               NaN   
...

In [16]:
categorical_features = Features.columns[Features.dtypes == "object"]

categorical_features = Features[categorical_features]
categorical_features

age_group         education      race     sex  \
respondent_id                                                      
0              55 - 64 Years        < 12 Years     White  Female   
1              35 - 44 Years          12 Years     White    Male   
2              18 - 34 Years  College Graduate     White    Male   
3                  65+ Years          12 Years     White  Female   
4              45 - 54 Years      Some College     White  Female   
...                      ...               ...       ...     ...   
26702              65+ Years      Some College     White  Female   
26703          18 - 34 Years  College Graduate     White    Male   
26704          55 - 64 Years      Some College     White  Female   
26705          18 - 34 Years      Some College  Hispanic  Female   
26706              65+ Years      Some College     White    Male   

                          income_poverty marital_status rent_or_own  \
respondent_id                                                         
0                          Below Poverty    Not Married         Own   
1                          Below Poverty    Not Married        Rent   
2              <= $75,000, Above Poverty    Not Married         Own   
3                          Below Poverty    Not Married        Rent   
4              <= $75,000, Above Poverty        Married         Own   
...                                  ...            ...         ...   
26702          <= $75,000, Above Poverty    Not Married         Own   
26703          <= $75,000, Above Poverty    Not Married        Rent   
26704                                NaN    Not Married         Own   
26705          <= $75,000, Above Poverty        Married        Rent   
26706          <= $75,000, Above Poverty        Married         Own   

                employment_status hhs_geo_region                census_msa  \
respondent_id                                                                
0              Not in Labor Force       oxchjgsf                   Non-MSA   
1                        Employed       bhuqouqj  MSA, Not Principle  City   
2                        Employed       qufhixun  MSA, Not Principle  City   
3              Not in Labor Force       lrircsnp       MSA, Principle City   
4                        Employed       qufhixun  MSA, Not Principle  City   
...                           ...            ...                       ...   
26702          Not in Labor Force       qufhixun                   Non-MSA   
26703                    Employed       lzgpxyit       MSA, Principle City   
26704                         NaN       lzgpxyit  MSA, Not Principle  City   
26705                    Employed       lrircsnp                   Non-MSA   
26706          Not in Labor Force       mlyzmhmf       MSA, Principle City   

              employment_industry employment_occupation  
respondent_id                                            
0                             NaN                   NaN  
1                        pxcmvdjn              xgwztkwe  
2                        rucpziij              xtkaffoo  
3                             NaN                   NaN  
4                        wxleyezf              emcorrxb  
...                           ...                   ...  
26702                         NaN                   NaN  
26703                    fcxhlnwr              cmhcxjea  
26704                         NaN                   NaN  
26705                    fcxhlnwr              haliazsg  
26706                         NaN                   NaN  

[26707 rows x 12 columns]

### Dealing with the Null Entries
In this section, i will find suitable ways to deal with the null entries in the numeric_features and categorical_features dataframes.

For the numeric features, i will fill the null entries with their respective column median values.

In [17]:
# check for null entries
numeric_features.isna().sum()

h1n1_concern                      92
h1n1_knowledge                   116
behavioral_antiviral_meds         71
behavioral_avoidance             208
behavioral_face_mask              19
behavioral_wash_hands             42
behavioral_large_gatherings       87
behavioral_outside_home           82
behavioral_touch_face            128
doctor_recc_h1n1                2160
doctor_recc_seasonal            2160
chronic_med_condition            971
child_under_6_months             820
health_worker                    804
health_insurance               12274
opinion_h1n1_vacc_effective      391
opinion_h1n1_risk                388
opinion_h1n1_sick_from_vacc      395
opinion_seas_vacc_effective      462
opinion_seas_risk                514
opinion_seas_sick_from_vacc      537
household_adults                 249
household_children               249
dtype: int64

In [18]:
# fill the null entries with the respective column means
numeric_features = numeric_features.fillna(numeric_features.median())

In [19]:
# check for null entries
numeric_features.isna().sum()

h1n1_concern                   0
h1n1_knowledge                 0
behavioral_antiviral_meds      0
behavioral_avoidance           0
behavioral_face_mask           0
behavioral_wash_hands          0
behavioral_large_gatherings    0
behavioral_outside_home        0
behavioral_touch_face          0
doctor_recc_h1n1               0
doctor_recc_seasonal           0
chronic_med_condition          0
child_under_6_months           0
health_worker                  0
health_insurance               0
opinion_h1n1_vacc_effective    0
opinion_h1n1_risk              0
opinion_h1n1_sick_from_vacc    0
opinion_seas_vacc_effective    0
opinion_seas_risk              0
opinion_seas_sick_from_vacc    0
household_adults               0
household_children             0
dtype: int64

From the output above, we have successfully filled our missing values with their respective column means.

Next, for the categorical features, i will fill the missing values with unknown.

In [20]:
# check for null entries
categorical_features.isna().sum()

age_group                    0
education                 1407
race                         0
sex                          0
income_poverty            4423
marital_status            1408
rent_or_own               2042
employment_status         1463
hhs_geo_region               0
census_msa                   0
employment_industry      13330
employment_occupation    13470
dtype: int64

In [21]:
# fill the null entries with unknown
categorical_features = categorical_features.fillna('unknown')

In [22]:
# check for null entries
categorical_features.isna().sum()

age_group                0
education                0
race                     0
sex                      0
income_poverty           0
marital_status           0
rent_or_own              0
employment_status        0
hhs_geo_region           0
census_msa               0
employment_industry      0
employment_occupation    0
dtype: int64

Based on the output above, we have successfully dealt with the caegorical features missing values.

Now using the cleaned numeric_features and categorical_features dataframes, i will mearge the two to create a new dataframe containing the cleaned data

In [31]:
cleaned_Features= pd.merge(numeric_features, categorical_features, on="respondent_id", how='inner')
cleaned_Features

h1n1_concern  h1n1_knowledge  behavioral_antiviral_meds  \
respondent_id                                                            
0                       1.0             0.0                        0.0   
1                       3.0             2.0                        0.0   
2                       1.0             1.0                        0.0   
3                       1.0             1.0                        0.0   
4                       2.0             1.0                        0.0   
...                     ...             ...                        ...   
26702                   2.0             0.0                        0.0   
26703                   1.0             2.0                        0.0   
26704                   2.0             2.0                        0.0   
26705                   1.0             1.0                        0.0   
26706                   0.0             0.0                        0.0   

               behavioral_avoidance  behavioral_face_mask  \
respondent_id                                               
0                               0.0                   0.0   
1                               1.0                   0.0   
2                               1.0                   0.0   
3                               1.0                   0.0   
4                               1.0                   0.0   
...                             ...                   ...   
26702                           1.0                   0.0   
26703                           1.0                   0.0   
26704                           1.0                   1.0   
26705                           0.0                   0.0   
26706                           1.0                   0.0   

               behavioral_wash_hands  behavioral_large_gatherings  \
respondent_id                                                       
0                                0.0                          0.0   
1                                1.0                          0.0   
2                                0.0                          0.0   
3                                1.0                          1.0   
4                                1.0                          1.0   
...                              ...                          ...   
26702                            0.0                          0.0   
26703                            1.0                          0.0   
26704                            1.0                          1.0   
26705                            0.0                          0.0   
26706                            0.0                          0.0   

               behavioral_outside_home  behavioral_touch_face  \
respondent_id                                                   
0                                  1.0                    1.0   
1                                  1.0                    1.0   
2                                  0.0                    0.0   
3                                  0.0                    0.0   
4                                  0.0                    1.0   
...                                ...                    ...   
26702                              1.0                    0.0   
26703                              0.0                    0.0   
26704                              0.0                    1.0   
26705                              0.0                    1.0   
26706                              0.0                    0.0   

               doctor_recc_h1n1  ...      race     sex  \
respondent_id                    ...                     
0                           0.0  ...     White  Female   
1                           0.0  ...     White    Male   
2                           0.0  ...     White    Male   
3                           0.0  ...     White  Female   
4                           0.0  ...     White  Female   
...                         ...  ...       ...     ...   
26702                       0.0  ...     White  Fe

## Checking For Duplicates

In this section, i will checked both the "cleaned_features" and "target" dataframes for duplicates and find appropriate ways of tackling them.

In [32]:
# check for duplicates in the cleaned_features dataframe
cleaned_Features.duplicated().sum()

1

Based on the output, we have only one duplicated values. I will drop the duplicated entry.

In [33]:
# Drop duplicates
cleaned_Features.drop_duplicates(inplace=True)

In [41]:
# check the target dataframe for duplicates
Target.duplicated().sum()

26705

Based on the output there are about 26705 duplicated values in the target dataframe. This makes sense since the target dataframe is categorical and contains two class label, either 1 or 0. Therefore, i will not drop the duplicated values.

## Checking for Outliers